include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,205 @@
 #ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
-#endif
-
+				CASE_BEGIN(ndf, NDF_GGX) {
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(s, microfacet.isotropic, currInteraction.isotropic, a2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
-
-
-#ifndef ONLY_ONE_NDF
-		} else
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
 #endif
+				} CASE_END
 #endif
 
 #ifdef NDF_BECKMANN
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_BECKMANN) {
-#endif
-
+				CASE_BEGIN(ndf, NDF_BECKMANN) {
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
+					bxdf_eval_scalar_part = irr_glsl_beckmann_height_correlated_cos_eval_DG(s, microfacet.isotropic, currInteraction.isotropic, a2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_DG(currBSDFParams, currInteraction, a, a2, ay, ay2);
-#endif
-
-#ifndef ONLY_ONE_NDF
-		} else
+					bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
 #endif
+				} CASE_END
 #endif
 
 #ifdef NDF_PHONG
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_PHONG) {
-#endif
-
-
-			float n = irr_glsl_alpha2_to_phong_exp(a2);
+				CASE_BEGIN(ndf, NDF_PHONG) {
+					float n = irr_glsl_alpha2_to_phong_exp(a2);
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, n, a2);
+					bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(s, microfacet.isotropic, currInteraction.isotropic, n, a2);
 #else
-			float ny = irr_glsl_alpha2_to_phong_exp(ay2);
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
+					float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+					bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(s, microfacet, currInteraction, n, ny, a2, ay2);
+#endif
+				} CASE_END
 #endif
 
+				CASE_OTHERWISE
+				{} //else "empty braces"
+			END_CASES
 
-#ifndef ONLY_ONE_NDF
-		} else
-#endif
+				float VdotH = microfacet.isotropic.VdotH;
+				float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
+				vec3 fr;
+#ifdef OP_CONDUCTOR
+				if (op == OP_CONDUCTOR)
+					fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+				else
 #endif
+					fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
 
-#ifndef ONLY_ONE_NDF
-		{} //else "empty braces"
-#endif
-	}
+				const float NdotL = cosFactor;
+				const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
```

Comment:
move the computation of clamped `NdotV` outside (initialize `is_valid` with `NdotV>FLT_MIN`)

the clamped `VdotH` can also be moved up and used in the valid microfacet check somewhat (but before clamp, to check that `VdotH>=0.0`)

Reply:
never mind the `VdotH` check, i forgot this is EVAL only